### PR TITLE
Allow AssetsConfig and OptimizelyConfig to be passed into templates

### PR DESCRIPTION
**Pull request by frankoid**
This allows controllers that use these templates to be unit tested without a running Play application, by passing in test AssetsConfig and OptimizelyConfig implementations that do not use the Play config. For an example of how to do this see https://github.com/hmrc/agent-subscription-frontend/blob/no-play-app-in-unit-test/test/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionControllerSpec.scala

N.B. I think a new major version will be needed after this change because I don't think it's binary compatible. It adds new arguments to `footer.scala.html` and `head.scala.html`, although they do have default values so most users of the library will just need to be recompiled.

### CODE REVIEW

File: src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@import uk.gov.hmrc.play.config.AssetsConfig
+@()(implicit assetsConfig: AssetsConfig, optimizelyConfig: OptimizelyConfig)
+
+@*
+The HTML produced by this template doesn't need to make sense, it just
+needs to include the templates which use configuration to ensure that they
+can be rendered in tests without a running Play application (by passing in
+test AssetsConfig and OptimizelyConfig instances that do not access Play
+configuration).
+*@
```
Comment (feedmypixel) [inline src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html]: This could be made more concise:

```The HTML produced by this template just needs to include the templates which use configuration to ensure that they can be rendered in tests without a running Play application```

File: src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (feedmypixel) [inline src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: This `AkkaMaterializerSpec` feels like a testing preference, if thats the case it should be a separate PR and added in for all tests.

File: src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
+@import uk.gov.hmrc.play.config.OptimizelyConfig
```
Comment (feedmypixel) [inline src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html]: IMO it feels like the name of the file `test_includes_that_use_config.scala.html` isn't very clear and could just be `test_header_footer_includes.scala.html` and then the comment in the file explains the particulars

File: src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+class TestWithoutPlayAppSpec extends WordSpec with Matchers with FutureAwaits with DefaultAwaitTimeout with AkkaMaterializerSpec {
+
+  override implicit val defaultAwaitTimeout: Timeout = 5 seconds
+
+  "templates that use Play configuration (head and footer)" should {
```
Comment (feedmypixel) [inline src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala]: You could be explicit - `Header and Footer templates that use Play configuration`

**Review by feedmypixel — COMMENTED** [review]
@frankoid thanks for the contribution. I like this approach and agree this needs to be a major.
I have added a few comments. Thanks

File: src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html (hunk 3 above)
Comment (frankoid) [reply src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html]: agreed, will change

File: src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html (hunk 1 above)
Comment (frankoid) [reply src/test/twirl/uk/gov/hmrc/play/views/layouts/test_includes_that_use_config.scala.html]: agreed, will change

File: src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala (hunk 2 above)
Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: no, AkkaMaterializerSpec is needed by the new test but not any other tests. Try removing `with AkkaMaterializerSpec` from `TestWithoutPlayAppSpec` and then running `TestWithoutPlayAppSpec` - it will fail. Whereas all the other tests run fine without it

File: src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala (hunk 4 above)
Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala]: agreed, will change

File: src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala (hunk 2 above)
Comment (feedmypixel) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: Is there a benefit to using `AkkaMaterializerSpec` rather than [uk.gov.hmrc.play.test.UnitSpec](https://github.com/hmrc/hmrctest/blob/master/src/main/scala/uk/gov/hmrc/play/test/UnitSpec.scala)?

Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: Yes. Whether `AkkaMaterializerSpec` is needed is independent of whether `UnitSpec` is used.

`AkkaMaterializerSpec`'s job is to provide an implicit Materializer. An implicit Materializer is needed by `bodyOf` (whichever version of `bodyOf` you use - `TestWithoutPlayAppSpec.bodyOf` and `UnitSpec.bodyOf` both need one).

The reason I didn't use `UnitSpec` is that the other specs in this library don't use it and it's not on the classpath. In doing so I actually discovered that I like the traits in `play.api.test` better than UnitSpec because they are more composable. However I don't mind adding a dependency on `hmrctest` and making `TestWithoutPlayAppSpec` use `UnitSpec` if that's what you prefer - but that won't remove the need for `AkkaMaterializerSpec`.

The only way I can think of of getting rid of `AkkaMaterializerSpec` is to merge it into `TestWithoutPlayAppSpec` (which is the only test that needs it so far).

Aside: `UnitSpec.bodyOf` has only needed a Materializer since the Play 2.5 upgrade.

File: src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala (hunk 4 above)
Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/play/views/layouts/TestWithoutPlayAppSpec.scala]: just a note on the intention of this code - the intention is that if somebody makes another template need configuration they should add it to this test, to make sure that all templates can be rendered without a started Play app

so another way to word this (for explanation, I'm not suggesting we actually put this in the code) is "templates that use Play configuration, which at the time of writing happen to be head and footer but which could include other templates in future"

**Comment by frankoid** [comment]
@feedmypixel changes pushed, I can squash the commits if you like?

File: src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala (hunk 2 above)
Comment (feedmypixel) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: What I am trying to understand is why `AkkaMaterializerSpec` has been included? Can you not just call the html fragments in the tests directly and override the`assetsConfig` and `optimizelyConfig` implicits?
```
@head(
linkElem = None, 
headScripts = Some(headScripts))
(assetsConfig, optimizelyConfig)
```
```
@footer(
 analyticsToken = None,
 analyticsHost = "",
 ssoUrl = None,
 scriptElem = Some(footerScriptElem),
 gaCalls = None)
(assetsConfig, optimizelyConfig)

```
 In the same fashion as [InputRadioGroupSpec](https://github.com/hmrc/play-ui/blob/master/src/test/scala/uk/gov/hmrc/play/views/helpers/InputRadioGroupSpec.scala#L45)

Then if I have understood correctly there would be no need for the `TestWithoutPlayAppController.scala`, `AkkaMaterializerSpec.scala` or the `test_header_footer_includes.scala.html`

**Comment by feedmypixel** [comment]
> @feedmypixel changes pushed, I can squash the commits if you like?

Thanks

File: src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala (hunk 2 above)
Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: ah I see. Perhaps the Materializer is only needed because I'm testing the templates via a controller? I will see what happens if I get rid of the controller and call the templates directly from the test like InputRadioGroupSpec does

Comment (frankoid) [reply src/test/scala/uk/gov/hmrc/test/AkkaMaterializerSpec.scala]: yes that works, much simpler now :)

**Comment by feedmypixel** [comment]
@frankoid Thank you for the contribution. I will release `v7.0`
